### ISSUE TITLE
How to list and include all available linters?

### ISSUE THREAD

**Issue by IndrajeetPatil** [r-lib/lintr#1482]
# Synopsis

Correct me if I missed this, but I don't think currently the docs/functions make it easy to:

- list names of all available linters
- include all available linters while linting (a package, e.g.)

# Listing all available linters

It should be possible for the users to easily generate a list of names of available linters. 

Currently, we make this easy only for default linters:

``` r
library(lintr)

names(default_linters) # or `names(default_settings$linters)` will also work
#>  [1] "assignment_linter"                "brace_linter"                    
#>  [3] "commas_linter"                    "commented_code_linter"           
#>  [5] "cyclocomp_linter"                 "equals_na_linter"                
#>  [7] "function_left_parentheses_linter" "infix_spaces_linter"             
#>  [9] "line_length_linter"               "no_tab_linter"                   
#> [11] "object_length_linter"             "object_name_linter"              
#> [13] "object_usage_linter"              "paren_body_linter"               
#> [15] "pipe_continuation_linter"         "semicolon_linter"                
#> [17] "seq_linter"                       "single_quotes_linter"            
#> [19] "spaces_inside_linter"             "spaces_left_parentheses_linter"  
#> [21] "T_and_F_symbol_linter"            "trailing_blank_lines_linter"     
#> [23] "trailing_whitespace_linter"       "vector_logic_linter"
```

<sup>Created on 2022-07-28 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

But we have many other linters, which are not included in `default_linters` (why?!), and can be seen only* in the [reference](https://lintr.r-lib.org/reference/index.html#individual-linters).

This means that if I wanted to include all available linters, I need to manually inspect the list of default linters and the website, note that down which are not included and then add them to my `.lintr` file. That is cumbersome.

(* There is `lintr:::rd_linterlist()` to list all available linters, but it is not in a form that is useful, and it is not exported.)

So should we implement and export a new `all_linters` list, similar to `default_linters`?

# Linting a package with all available linters

If we implement and export `all_linters`, it also solves the second issue, which is to lint a package will all available linters. Because a user can just do:

```r
with_defaults(..., default = all_linters)
```

I also think that this should be the default, but I am guessing there are good reasons why only certain - and not available - linters were chosen for the default list.

**Comment by MichaelChirico** [comment]
This is what `available_linters(tags = NULL)` is for IINM.

We don't include all linters in `default_linters` because `default_linters` is targeted for the tidyverse style guide (this is not strictly true, but basically so). Many linters are helpful, but don't specifically enforce style guide rules, or have many known edge cases (`absolute_path_linter` comes to mind), or wouldn't always be helpful (e.g., the `expect_*` testthat-specific linters aren't useful to maintainers not relying on testthat).

**Comment by MichaelChirico** [comment]
Oh, and to use these, you'd use `linters_with_tags(tags = NULL)` to activate all linters

**Comment by IndrajeetPatil** [comment]
Thanks for the detailed reply!

I can confirm that the following activates all available linters:

```r
lintr::lint_package(linters = lintr::linters_with_tags(tags = NULL))
```

Sorry, missed this in the docs, but IMO we should include this front and center in the `README`.

I ran it with one of my packages, and found so many new lints! 
This is why I wish we would highlight the option of using all available linters more vigorously. 

If you agree, this is now just a documentation issue.

**Comment by MichaelChirico** [comment]
marked it as a documentation issue -- if a regular contributor is only now learning, we need to advertise it better!
